pkg/storage/gcp/bucket_mock.go

```diff
@@ -0,0 +1,95 @@
+package gcp
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"strings"
+	"sync"
+)
+
+type bucketMock struct {
+	db   map[string][]byte
+	lock sync.RWMutex
+}
+
+func newBucketMock() Bucket {
+	b := &bucketMock{}
+	b.db = make(map[string][]byte)
+	return b
+}
+
+type bucketReader struct {
+	io.Reader
+	*bucketMock
+}
+
+type bucketWriter struct {
+	*bucketMock
+	path string
+}
+
+func (r *bucketReader) Close() error {
+	r.bucketMock.lock.RUnlock()
+	return nil
+}
+
+func (r *bucketWriter) Close() error {
+	r.bucketMock.lock.Unlock()
```

Comment:
can you help me understand why you need the unlock here and above?

Reply:
It might sound an overkill since the test are all called sequentially (and there are probably better ways to achieve the same result), however what I wanted to achieve was to protect the db until the whole read / write was done, treating it as a resource. I took leverage of the fact that Read and Write return a io.ReadCloser / io.WriteCloser and that the user of the api is expected to release the resource by closing the reader / writer.
So the mutex is taken while returning the reader / writer and is released while closing the reader / writer.

Mutexing the single Read / Write call would not be enough if the payload is consumed in more than one read / write call. Consider the case of a read that reads one byte at the time while a writer changes the record from a different goroutine for example.

Reply:
While we're at it - we could record the call to Close() (setting a bool is probably enough) and check it later in a test so that we know that it was called. Just a thought. What do you think @fedepaol ?

Reply:
@fedepaol thank you for that amazing response and explanation. that lock-unlock flow you described sounds great to me! Also, +1 on the counter that @marpio mentioned 👍